torch/utils/_python_dispatch.py

```diff
@@ -173,6 +86,56 @@ def push(cls, *args, **kwargs):
         return instance
 
 
+# WARNING: If you access this list to see all the active modes in order, do not update the list in place
+# since it alters the actual mode stack
+threadLocal = threading.local()
+def _get_cur_mode_stack():
+    mode_stack = getattr(threadLocal, 'dispatch_mode_stack', None)
+    if not mode_stack:
+        mode_stack = []
+        threadLocal.dispatch_mode_stack = mode_stack
+    return mode_stack
+
+
+def get_current_dispatch_mode():
+    return _get_cur_mode_stack()[-1] if len(_get_cur_mode_stack()) > 0 else None
+
+
+def _push_mode(mode):
+    if len(_get_cur_mode_stack()) == 0:
+        _set_torch_dispatch_mode(_TorchDispatchStackMode())
+    _get_cur_mode_stack().append(mode)
+
+
+def _pop_mode():
+    assert len(_get_cur_mode_stack()) > 0
+    old = _get_cur_mode_stack().pop()
+    if len(_get_cur_mode_stack()) == 0:
+        _set_torch_dispatch_mode(None)
+    return old
+
+
+@contextlib.contextmanager
+def _pop_mode_temporarily():
+    old = _pop_mode()
+    try:
+        yield old
+    finally:
+        _push_mode(old)
+
+# a helper "mode" used by the torch dispatch push helper method. This is the only mode that will ever
+# be active at the C++ level and it will run the current mode
+class _TorchDispatchStackMode:
+    def __torch_dispatch__(self, func, types, args=(), kwargs=None):
+        with _pop_mode_temporarily() as old:
+            if len(_get_cur_mode_stack()) > 0:
+                _set_torch_dispatch_mode(self)
+            # we can't check the type of __torch_function__ here but this is sufficient for checking it's a classmethod
+            if old.__torch_dispatch__.__self__ is type(old):
+                raise RuntimeError("TorchFunctionMode's torch_function function " +
+                                   "should be a normal method not a class method")
+            return old.__torch_dispatch__(func, types, args, kwargs)
```

Comment:
- Should the error message say TorchDispatchMode?
- Why is `old.__torch_dispatch__.__self__ is type(old)` the correct check here?